### PR TITLE
dxf/export: register export task type and add the dump subtask generation

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: ref #68984

Problem Summary:

First half of the distributed export DXF backend (design M2a): register the `Export` task type, split the `Dump` step into chunks, and size the task in an `OnPrepare` step. Part of the `EXPORT TABLE` / `EXPORT SCHEMA` feature tracked in #68984.

### What changed and how does it work?

Table-set-native from the start, so if we want to add `EXPORT SCHEMA` later, they can share this backend and implementation; single-table export is just `len(Tables) == 1`.

- **proto**: register the `Export` task type and the `ExportStepDump` step.
- **pkg/dxf/export** (new):
  - **Chunk model (borrowed from `IMPORT INTO`)** — EXPORT carves each table's key space into fixed-size (`chunkSize`) **chunks** and packs those into **subtasks**.
  - **Idempotent output** — a chunk's table-local `Ordinal`, fixed when the chunk is created, determines its output file names, so any concurrency, retry, or rebalancing reproduces the same files regardless of the worker count.
  - **Prepare-mode sizing** — like IMPORT INTO's prepare mode, an `OnPrepare` step runs after submission to estimate the data volume and derive the task's resources, keeping that heavy work off the submitting session.
- **pkg/store/helper, pkg/store/copr, pkg/ddl**: extract the region-load and per-region size helpers so add-index backfill and export share one implementation.
- **autoscaler / session**: add the export resource-sizing variant and register the Export scheduler.

The subtask executor (read → encode → upload pipeline) and the GC safepoint keeper are added by the next PR (M2b).

**Inert**: no `Export` task is ever submitted (no executor yet, no parser grammar, and `OnPrepare` stays dormant until the submission path sets `ExtraParams.PrepareMode`), so this is unreachable from SQL and existing behavior is unchanged.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

> Added `pkg/dxf/export` unit tests: chunk cutting and the fallback split, subtask packing, the physical-table record-key range (int and common handle), physical-id enumeration across partitions, and the task key. The full region-loading split and `OnPrepare` are exercised by the realtikv e2e in a later milestone.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/dxf/export/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	return nil
+}
+
+// GetEligibleInstances implements scheduler.Extension.
+func (*exportScheduler) GetEligibleInstances(context.Context, *proto.Task) ([]string, error) {
+	return nil, nil
+}
+
+// IsRetryableErr implements scheduler.Extension.
+func (*exportScheduler) IsRetryableErr(error) bool {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go]: Intentional for this milestone — this mirrors `LitBackfillScheduler.IsRetryableErr`, which is also unconditionally true. Real error classification (revert on non-transient config/type errors, keep retrying transient PD/region ones) will land together with the Dump executor in a later PR, where there's an actual running task to revert. Latent until then since no Export task is submitted yet.

File: pkg/dxf/export/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	return true
+}
+
+// ModifyMeta implements scheduler.Extension.
+func (*exportScheduler) ModifyMeta(oldMeta []byte, _ []proto.Modification) ([]byte, error) {
+	return oldMeta, nil
+}
+
+// TaskKey returns the DXF task key from the root table/schema id and snapshot.
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go]: Done in https://github.com/pingcap/tidb/commit/ccece2cc5a — renamed the param to `rootID` and spelled out the table-id (EXPORT TABLE) vs schema-id (EXPORT SCHEMA) meaning in the doc.

File: pkg/dxf/export/split.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+	// maxChunksPerSubtask is a hard ceiling on the chunk count per subtask (the
+	// usual bound is byte size), so a schema of many tiny tables stays bounded.
+	maxChunksPerSubtask = 4000
+)
+
+// splitTables carves every table into ~chunkSize chunks, then packs the chunks
+// into subtasks sized to the worker concurrency.
+func splitTables(ctx context.Context, store kv.Storage, meta *TaskMeta, concurrency int) ([][]byte, error) {
+	chunks := make([]Chunk, 0, len(meta.Tables))
+	for tableIdx := range meta.Tables {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Known trade-off — this is the per-partition PD-scan cost for EXPORT SCHEMA, already on the follow-up list (parallelize the per-physical scans). The near-term target is single-table export, which only walks one table's partitions, so leaving it serial for this PR.

File: pkg/dxf/export/split.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+			}
+			tableChunks, ordinal = chunksByCount(tableIdx, pid, boundaries, meta.PhysicalSizes[pid], ordinal)
+		}
+		chunks = append(chunks, tableChunks...)
+	}
+	return chunks, nil
+}
+
+// regionChunks returns each region's end key and byte size over [start, end) from
+// PD, retried. ok is false when PD is unavailable.
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Done in https://github.com/pingcap/tidb/commit/ccece2cc5a — the comment now enumerates all four fallback conditions (no region cache, PD unavailable, retries exhausted, empty result).

File: pkg/dxf/export/split.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+func regionChunks(ctx context.Context, store kv.Storage, start, end kv.Key) (endKeys []kv.Key, sizes []int64, ok bool) {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Agreed — renamed to `loadRegionSizes` in https://github.com/pingcap/tidb/commit/ccece2cc5a, since it returns per-region sizes rather than Chunks.

File: pkg/dxf/export/split.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+	}
+	h := helper.NewHelper(hStore)
+	pdCli, err := h.TryGetPDHTTPClient()
+	if err != nil {
+		return nil, nil, false
+	}
+	backoffer := backoff.NewExponential(scanRegionBackoffBase, 2, scanRegionBackoffMax)
+	err = handle.RunWithRetry(ctx, loadRegionMaxRetry, backoffer, logutil.BgLogger(), func(context.Context) (bool, error) {
+		endKeys, sizes, err = h.RegionApproximateSizes(ctx, pdCli, start, end)
+		return err != nil, err
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Done in https://github.com/pingcap/tidb/commit/ccece2cc5a — added a warning log when it falls back to the count-based split. Kept the retry-all behaviour though: the errors here all come from PD `GetRegionsByKeyRange`, which is the transient class worth retrying; the missing observability was the real gap.

File: pkg/dxf/export/split.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		ord++
+		chunkStart = chunkEnd
+		acc = 0
+	}
+	return chunks, ord
+}
+
+// chunksByCount is the fallback when per-region sizes are unavailable: equal
+// region-count groups, apportioning totalSize by region count.
+func chunksByCount(tableIdx int, pid int64, boundaries []kv.Key, totalSize int64, startOrdinal int) ([]Chunk, int) {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Leaving these as two builders for now — the `Chunk{...}`/ordinal threading is small and both paths are covered by `TestChunksBySize`/`TestChunksByCount`, so a shared `newChunk` helper felt like more indirection than it removes. Happy to revisit if they grow.

File: pkg/store/helper/helper.go [hunk 8]
```diff
@@ -849,6 +849,57 @@ func (h *Helper) GetPDRegionStats(ctx context.Context, tableID int64, noIndexSta
...
+		if bytes.Compare(cur, end) >= 0 {
+			break
+		}
+	}
+	return endKeys, sizes, nil
+}
+
+// EstimateKeyRangeSize sums RegionApproximateSizes over [startKey, endKey).
+func (h *Helper) EstimateKeyRangeSize(ctx context.Context, pdCli pd.Client, startKey, endKey kv.Key) (int64, error) {
+	_, sizes, err := h.RegionApproximateSizes(ctx, pdCli, startKey, endKey)
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/store/helper/helper.go]: Done in https://github.com/pingcap/tidb/commit/ccece2cc5a — `EstimateKeyRangeSize` keeps its own O(1) sum loop again (decodes only the page cursor), so the reorg path no longer materializes per-region end keys. The decoding `RegionApproximateSizes` stays for the export split path that actually needs the keys.

**Comment by joechenrh** [comment]
On the three summary-only (unanchored) findings:

1. **Shared retry helper for region loading** — keeping ddl and dxf/export separate deliberately. The retry wrapper uses `handle.RunWithRetry` from `pkg/dxf/framework/handle`, which cannot move into `pkg/store/copr` without an import cycle. And backfill's loop is not the same shape as export's: it injects the `mockPhysicalTableRegionDiscontinuity` failpoint between the load and the continuity check, and wraps per-batch TS allocation inside the retry. The genuinely shared piece (sort + continuity check) already lives in `copr.LoadSortedContinuousRegions`.
2. **chunkSize/FileSize invariant** — softened the comment in ccece2cc5a to state it is an unenforced design assumption.
3. **chunksPerWorker naming** — clarified in ccece2cc5a that it is a byte-budget multiplier (`chunksPerWorker*concurrency*chunkSize`), not a literal per-subtask chunk count.

File: pkg/dxf/export/scheduler.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	s.BaseScheduler.Extension = s
+	return s.BaseScheduler.Init()
+}
+
+// OnTick implements scheduler.Extension.
+func (*exportScheduler) OnTick(context.Context, *proto.Task) {}
+
+// OnPrepare implements scheduler.Extension. In prepare mode it estimates the
+// export set's data size once the task is submitted, seeds the per-physical sizes
+// used by the split, and sizes the task's resources from the total — keeping the
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go]: Done in https://github.com/pingcap/tidb/commit/622fe14479 — the doc now notes resource sizing is nextgen-only.

File: pkg/store/helper/helper.go [hunk 10]
```diff
@@ -849,6 +849,72 @@ func (h *Helper) GetPDRegionStats(ctx context.Context, tableID int64, noIndexSta
...
+			break
+		}
+	}
+	return endKeys, sizes, nil
+}
+
+// EstimateKeyRangeSize sums each region's max(ApproximateSize, ApproximateKvSize)
+// over [startKey, endKey) via PD. Unlike RegionApproximateSizes it keeps O(1)
+// memory and decodes only the page cursor, since callers here need just the total.
+func (h *Helper) EstimateKeyRangeSize(ctx context.Context, pdCli pd.Client, startKey, endKey kv.Key) (int64, error) {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/store/helper/helper.go:925]: Intentional: last round's finding (helper.go:892) asked me to give `EstimateKeyRangeSize` its own O(1) loop so the DDL reorg path stops materializing per-region end keys. Sharing the pagination again would reintroduce exactly that regression — `RegionApproximateSizes` must retain per-region keys, `EstimateKeyRangeSize` only sums, so the two genuinely want different loops.

File: pkg/ddl/backfilling_dist_scheduler.go [hunk 11]
```diff
@@ -356,31 +357,18 @@ func generatePlanForPhysicalTable(
...
-		}
-		sort.Slice(recordRegionMetas, func(i, j int) bool {
-			return bytes.Compare(recordRegionMetas[i].StartKey(), recordRegionMetas[j].StartKey()) < 0
+		recordRegionMetas, err := copr.LoadSortedContinuousRegions(
+			tikv.NewBackofferWithVars(context.Background(), 20000, nil), regionCache, startKey, endKey)
+		failpoint.Inject("mockPhysicalTableRegionDiscontinuity", func() {
+			err = copr.ErrRegionsNotContinuous
 		})
-
 		// LoadRegionsInKeyRange can combine multiple PD scans. A concurrent region
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/ddl/backfilling_dist_scheduler.go]: Done in https://github.com/pingcap/tidb/commit/622fe14479 — dropped the stale `LoadRegionsInKeyRange` reference; the retry rationale now lives on `copr.LoadSortedContinuousRegions`.

File: pkg/dxf/export/proto.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+	SnapshotTS uint64      `json:"snapshot_ts"`
+	// PhysicalSizes maps a physical table (or partition) id to its estimated byte
+	// size, computed once at submit time (where it also feeds resource scaling).
+	// The split prefers a fresh estimate and falls back to this only when PD is
+	// unavailable.
+	PhysicalSizes map[int64]int64 `json:"physical_sizes"`
+	// Dest is the destination URI, with credentials in the query part.
+	Dest     string `json:"dest"`
+	Format   string `json:"format"`
+	FileSize int64  `json:"file_size"`
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/proto.go]: Done in https://github.com/pingcap/tidb/commit/622fe14479 — documented `Format` and `FileSize`.

File: pkg/dxf/export/split.go [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+	"github.com/pingcap/tidb/pkg/util/mathutil"
+	"github.com/tikv/client-go/v2/tikv"
+	"go.uber.org/zap"
+)
+
+const (
+	loadRegionMaxRetry    = 8
+	scanRegionBackoffBase = 200 * time.Millisecond
+	scanRegionBackoffMax  = 2 * time.Second
+	// chunkSize is the per-chunk work granularity. Assumed (not enforced against
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: `FileSize` is consumed by the subtask executor's file writer in the next PR (M2b), which cuts output CSV files at `FileSize` — it's carried in the meta here but read there, not dead. Its validation against chunkSize belongs where the EXPORT SQL wires FileSize up, which lands with that path; keeping it a soft assumption for now.

File: pkg/dxf/export/split.go [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+	sizes := make(map[int64]int64)
+	var total int64
+	for i := range meta.Tables {
+		tblInfo := meta.Tables[i].TableInfo
+		for _, pid := range physicalIDs(tblInfo) {
+			start, end := physicalTableRange(tblInfo, pid)
+			size, err := h.EstimateKeyRangeSize(ctx, pdCli, start, end)
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Done in https://github.com/pingcap/tidb/commit/622fe14479 — `estimateExportSize` now retries each table's PD estimate with the same backoff as the split-time scan, instead of aborting the whole prepare on a transient error.

File: pkg/dxf/export/split.go [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+	if tblInfo.IsCommonHandle {
+		return prefix, prefix.PrefixNext()
+	}
+	return tablecodec.EncodeRowKeyWithHandle(pid, kv.IntHandle(math.MinInt64)), prefix.PrefixNext()
+}
+
+// loadRegionBoundaries returns the sorted boundaries spanning [start, end] (both
+// included), retrying with backoff while the regions are not continuous.
+func loadRegionBoundaries(ctx context.Context, store kv.Storage, start, end kv.Key) ([]kv.Key, error) {
+	hStore, ok := store.(helper.Storage)
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Leaving the three assertions inline — they're trivial and the three sites diverge in what they need next (pdCli vs region cache), so a shared guard wouldn't save much. Can revisit if the file grows.

File: pkg/dxf/export/split.go [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+	if !ok {
+		return nil, errors.New("storage does not support region cache")
+	}
+	var boundaries []kv.Key
+	backoffer := backoff.NewExponential(scanRegionBackoffBase, 2, scanRegionBackoffMax)
+	err := handle.RunWithRetry(ctx, loadRegionMaxRetry, backoffer, logutil.BgLogger(), func(context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Same call as the earlier region-loading finding: the retry wrapper uses `handle.RunWithRetry` from the DXF framework, which can't move into `pkg/store/copr` without an import cycle, and backfill's loop also injects the `mockPhysicalTableRegionDiscontinuity` failpoint and wraps TS allocation. The shared primitive (`copr.LoadSortedContinuousRegions`) is already extracted.

File: pkg/store/copr/region_cache.go [hunk 17]
```diff
@@ -52,6 +53,34 @@ const (
...
+// ErrRegionsNotContinuous is returned by LoadSortedContinuousRegions when the
+// loaded regions do not fully cover the range without a gap. It is retryable:
+// the caller can reload against a refreshed region cache.
+var ErrRegionsNotContinuous = errors.New("regions are not continuous")
+
+// LoadSortedContinuousRegions loads the regions covering [startKey, endKey) via
+// the region cache, sorted by start key, and verifies they are continuous — a
+// gap would silently drop rows. It returns ErrRegionsNotContinuous when the
+// loaded regions are empty or have a gap.
+func LoadSortedContinuousRegions(bo *tikv.Backoffer, regionCache *tikv.RegionCache, startKey, endKey []byte) ([]*tikv.Region, error) {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/store/copr/region_cache.go:62]: The gap detection is exercised by backfill's `mockPhysicalTableRegionDiscontinuity` failpoint test (`TestBackfillingSchedulerLocalMode`). A direct unit test would have to mock `regionCache.LoadRegionsInKeyRange`, which isn't easily constructible, so I'm keeping the indirect coverage rather than reshape the function for testability.

File: pkg/store/helper/helper.go [hunk 18]
```diff
@@ -849,6 +849,72 @@ func (h *Helper) GetPDRegionStats(ctx context.Context, tableID int64, noIndexSta
 	return pdCli.GetRegionStatusByKeyRange(ctx, pd.NewKeyRange(startKey, endKey), false)
 }
 
+// RegionApproximateSizes returns each region's raw end key and byte size
+// (max(ApproximateSize, ApproximateKvSize), preferring the logical KV size) over
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/store/helper/helper.go]: Done in https://github.com/pingcap/tidb/commit/622fe14479 — the comment now explains max(): the KV size best tracks logical data but can be 0 when TiKV doesn't report it, so max() falls back to the SST size.

File: pkg/dxf/export/split_test.go [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+	require.Equal(t, chunks, again)
+
+	// Ordinal continues across partitions of the same table.
+	more, next2 := chunksByCount(2, 200, keys("a", "z"), size, next)
+	require.Equal(t, 4, more[0].Ordinal)
+	require.Equal(t, int64(200), more[0].PhysicalID)
+	require.Equal(t, 5, next2)
+}
+
+func mustChunks(chunks []Chunk, _ int) []Chunk { return chunks }
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/dxf/export/split_test.go]: Done in https://github.com/pingcap/tidb/commit/622fe14479 — renamed to `chunksOnly` since it just drops the ordinal return and never panics.

File: pkg/dxf/export/scheduler.go [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+	return s.taskKSTaskMgr
+}
+
+func (s *exportScheduler) snapshotInfoSchema() (infoschema.InfoSchema, error) {
+	var is infoschema.InfoSchema
+	err := s.taskKSMgr().WithNewSession(func(se sessionctx.Context) error {
+		var err error
+		is, err = domain.GetDomain(se).GetSnapshotInfoSchema(s.taskMeta.SnapshotTS)
```
Comment (D3Hunter) [inline pkg/dxf/export/scheduler.go]: **[P1] Avoid resolving the domain from a cross-keyspace session.** In the NextGen path, `taskKSMgr()` builds a manager from `SysSessionPool()`. Cross-keyspace sessions created by that pool have no domain attached (`createSessionWithOpt(..., dom=nil, ...)`), so `domain.GetDomain(se)` can return nil and this dereference panics during both prepare and dump planning. Resolve the snapshot InfoSchema through the owning domain/keyspace API instead, and add a system-keyspace scheduler/user-keyspace task test.

File: pkg/store/helper/helper.go [hunk 21]
```diff
@@ -849,6 +850,75 @@ func (h *Helper) GetPDRegionStats(ctx context.Context, tableID int64, noIndexSta
...
+				return nil, nil, err
+			}
+			raw, err := codec.DecodeRegionKey(encoded)
+			if err != nil {
+				return nil, nil, err
+			}
+			endKeys = append(endKeys, raw)
+			sizes = append(sizes, max(r.ApproximateSize, r.ApproximateKvSize)*units.MiB)
+		}
+		cur, err = hex.DecodeString(regions.Regions[len(regions.Regions)-1].EndKey)
```
Comment (D3Hunter) [inline pkg/store/helper/helper.go]: **[P1] Stop pagination at an unbounded region end key.** PD represents the maximum region's `EndKey` as empty. `hex.DecodeString("")` returns an empty slice, which compares less than a finite `end`, so the next request restarts at the beginning and this loop can repeat until cancellation; API-v2 decoding above may fail on the same empty key first. Treat an empty last `EndKey` as positive infinity and stop. The duplicate cursor update in `EstimateKeyRangeSize` needs the same guard, plus a regression test with an empty terminal `EndKey`.

File: pkg/dxf/export/scheduler.go [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+
+// ModifyMeta implements scheduler.Extension.
+func (*exportScheduler) ModifyMeta(oldMeta []byte, _ []proto.Modification) ([]byte, error) {
+	return oldMeta, nil
+}
+
+// TaskKey returns the DXF task key from the root id (the table id) and the
+// snapshot.
+func TaskKey(rootID int64, snapshotTS uint64) string {
+	return fmt.Sprintf("export/%d/%d", rootID, snapshotTS)
```
Comment (D3Hunter) [inline pkg/dxf/export/scheduler.go:338]: **[P2] Scope the task key by keyspace.** Table IDs are only unique within a keyspace, but the DXF task table's `task_key` is globally unique and submission checks active/history tasks by key alone. Two tenants exporting the same table ID at the same snapshot can therefore collide and one receives `ErrTaskAlreadyExists`. Include the task keyspace in this key, following the NextGen IMPORT INTO task-key format.

**Review by D3Hunter — COMMENTED** [review]
Correctness review: five actionable findings.

File: pkg/dxf/export/scheduler.go [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+// GetEligibleInstances implements scheduler.Extension.
+func (*exportScheduler) GetEligibleInstances(context.Context, *proto.Task) ([]string, error) {
+	return nil, nil
+}
+
+// IsRetryableErr implements scheduler.Extension.
+func (*exportScheduler) IsRetryableErr(error) bool {
+	return true
```
Comment (D3Hunter) [inline pkg/dxf/export/scheduler.go]: **[P2] Do not retry every planning error forever.** The framework reverts a task from prepare/plan only when `IsRetryableErr` returns false. Returning true unconditionally makes permanent failures such as a missing table at the snapshot, an expired snapshot, invalid external-storage configuration, or unsupported storage retry indefinitely while retaining task resources. Classify known transient failures and return false for permanent/unknown errors.

File: pkg/dxf/export/scheduler.go [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+	store, err := extstore.GetGlobalExtStorage(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	stepStr := proto.Step2Str(proto.Export, step)
+	metas := make([][]byte, 0, len(groups))
+	for i, g := range groups {
+		sm := &SubtaskMeta{Chunks: g}
+		sm.ExternalPath = dxfutil.PlanMetaPath(taskID, stepStr, i+1)
+		if err := sm.WriteJSONToExternalStorage(ctx, store, sm); err != nil {
```
Comment (D3Hunter) [inline pkg/dxf/export/scheduler.go:278]: **[P2] Register cleanup for offloaded plan metadata.** This writes one external object per subtask and can also leave a partial prefix when a later write fails, but Export registers no framework cleaner. The scheduler manager treats a task type with no cleaner as already cleaned before moving it to history, and the existing global dump GC does not scan these numeric task paths. Add an Export cleaner that removes the task's external-meta prefix on success, failure, cancellation, and partial planning.

File: pkg/dxf/export/proto.go [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+	TableIDs []int64 `json:"table_ids"`
+}
+
+// TaskMeta is the task meta of an export task.
+type TaskMeta struct {
+	DBs        []DBSpec `json:"dbs"`
+	SnapshotTS uint64   `json:"snapshot_ts"`
+	// PhysicalSizes maps a physical table (or partition) id to its estimated size.
+	PhysicalSizes map[int64]int64 `json:"physical_sizes"`
+	Dest          string          `json:"dest"`
```
Comment (D3Hunter) [inline pkg/dxf/export/proto.go]: DestURI?

File: pkg/dxf/export/proto.go [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// Chunk is a ~chunkSize key range of one physical table. Its table-local Ordinal
+// fixes the output file name at split time, so retries produce the same files.
+type Chunk struct {
+	TableIdx   int    `json:"table_idx"`
+	PhysicalID int64  `json:"physical_id"`
+	Start      []byte `json:"start"`
+	End        []byte `json:"end"`
+	Size       int64  `json:"size"`
```
Comment (D3Hunter) [inline pkg/dxf/export/proto.go]: maybe comment what this size means, region/kv size or estimated export size?

File: pkg/dxf/export/proto.go [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+	Ordinal    int    `json:"ordinal"`
+}
+
+// SubtaskMeta is a batch of chunks.
+type SubtaskMeta struct {
+	dxfutil.BaseExternalMeta
+	Chunks []Chunk `json:"chunks" external:"true"`
```
Comment (D3Hunter) [inline pkg/dxf/export/proto.go:80]: since those external meta is removed on subtask end, maybe some statistics data such as chunk count, total size info inside the subtask meta too

File: pkg/dxf/export/proto.go [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+// DBSpec is a database and the ids of its tables to export.
+type DBSpec struct {
+	DBName   string  `json:"db_name"`
+	TableIDs []int64 `json:"table_ids"`
+}
+
+// TaskMeta is the task meta of an export task.
+type TaskMeta struct {
+	DBs        []DBSpec `json:"dbs"`
```
Comment (D3Hunter) [inline pkg/dxf/export/proto.go:28]: do we have a limit on this, such as for the 1 million table case

File: pkg/dxf/export/scheduler.go (hunk 20 above)
Comment (D3Hunter) [inline pkg/dxf/export/scheduler.go]: we don't load schema for task keyspace, only system tables are loaded, for user table they should load through meta api

File: pkg/dxf/export/split.go [hunk 29]
```diff
@@ -0,0 +1,308 @@
...
+				return nil, err
+			}
+			chunks = append(chunks, tableChunks...)
+			tableIdx++
+		}
+	}
+	return divideSubtasks(chunks, nodeCount), nil
+}
+
+func tableInfoByID(ctx context.Context, is infoschema.InfoSchema, id int64) (*model.TableInfo, error) {
```
Comment (D3Hunter) [inline pkg/dxf/export/split.go]: in nextgen system ks, we don't load schemas for user tables

File: pkg/dxf/export/scheduler.go (hunk 20 above)
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go]: Good catch — this whole cross-keyspace-session path is gone now. `snapshotTableInfos` reads the snapshot straight off `s.store` via `meta.NewReader`, so it never touches a session or `domain.GetDomain` at all.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 . Added `TestSnapshotTableInfosUsesTaskStore` to cover the user-keyspace path.

File: pkg/store/helper/helper.go (hunk 21 above)
Comment (joechenrh) [reply pkg/store/helper/helper.go]: You're right, thanks for calling this out precisely. Both `RegionApproximateSizes` and `EstimateKeyRangeSize` now treat an empty last `EndKey` as "scan complete" before trying to decode/advance the cursor, so pagination stops at the terminal region instead of looping.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 . Added `TestRegionSizePaginationStopsAtTerminalRegion`, covering both classic and keyspace codecs for both functions.

File: pkg/dxf/export/scheduler.go (hunk 22 above)
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go:338]: Fixed — added `TaskKeyInKeyspace`, which is what `TaskKey` now delegates to. It follows the NextGen IMPORT INTO format and prefixes the key with the task keyspace name so table IDs from different keyspaces at the same snapshot don't collide.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 .

File: pkg/dxf/export/scheduler.go (hunk 23 above)
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go]: Agreed, unconditional retry was too broad. `IsRetryableErr` now delegates to `isRetryablePlanningError`, which classifies gRPC codes (`DeadlineExceeded`, `ResourceExhausted`, `Aborted`, `Unavailable`, `DataLoss`) plus `common.IsRetryableError` as retryable, and treats everything else — missing table, expired snapshot, bad storage config, etc. — as permanent.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 . Added `TestIsRetryableErr` covering both buckets.

File: pkg/dxf/export/scheduler.go (hunk 24 above)
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go:278]: Added `exportCleaner`, registered via `scheduler.RegisterCleanerFactory(proto.Export, ...)`. It walks the task's whole external-storage prefix (`<taskID>/`) and deletes everything under it, so it covers the prepared-plan file and every per-subtask plan file in one shot regardless of how far planning got — no need to track individual object names.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 (plus a small simplification in https://github.com/pingcap/tidb/pull/70153/commits/3eb4352c235d5da7ddf34ba83efba77a50f54305 ). `TestExportCleaner` checks it's scoped to one task's prefix and is safe to retry after the files are already gone.

File: pkg/dxf/export/proto.go (hunk 25 above)
Comment (joechenrh) [reply pkg/dxf/export/proto.go]: Yep, renamed to `DestURI` 🙂

Done in https://github.com/pingcap/tidb/pull/70153/commits/b80672fdb93a1172e36b4c57fa6e1ccad40c844b .

File: pkg/dxf/export/proto.go (hunk 26 above)
Comment (joechenrh) [reply pkg/dxf/export/proto.go]: Added a comment: it's PD's estimated byte size of [Start, End), not an exact count.

Done in https://github.com/pingcap/tidb/pull/70153/commits/b80672fdb93a1172e36b4c57fa6e1ccad40c844b .

File: pkg/dxf/export/proto.go (hunk 27 above)
Comment (joechenrh) [reply pkg/dxf/export/proto.go:80]: Good point. Added `ChunkCount`/`TotalSize` to `SubtaskMeta` as internal (non-`external`) fields, so they stay in the framework row and are still readable after the external chunk file gets cleaned up, even though `Chunks` itself is gone.

Done in https://github.com/pingcap/tidb/pull/70153/commits/b80672fdb93a1172e36b4c57fa6e1ccad40c844b . `TestMarshalSubtasks` checks the row carries the summary but not the chunk payload.

File: pkg/dxf/export/proto.go (hunk 28 above)
Comment (joechenrh) [reply pkg/dxf/export/proto.go:28]: No hard limit yet, and I want to be upfront that this isn't fully solved. What we have: `TaskMeta.DBs` groups tables by database (`DBSpec{DBName, TableIDs}`) instead of repeating the db name per table, which helps the common multi-database case a lot (measured ~58% smaller at 100k tables / 25k dbs, e.g. 4.67MiB -> 1.96MiB). But it doesn't help a single huge database — 1M tables in one db still serializes to ~9.5MiB, over the 6MB txn-entry-size row limit.

The real fix is externalizing `TaskMeta.DBs` to S3 the same way `SubtaskMeta.Chunks` already is (`dxfutil.BaseExternalMeta` + `external:"true"`), hydrated in `OnPrepare` since `Init()` has no ctx to read from external storage. We're deliberately deferring that until the submission-side code (which actually constructs `TaskMeta.DBs`) lands, since designing the external layout without a real caller felt premature. Tracking it as a known follow-up rather than doing it speculatively now.

File: pkg/dxf/export/scheduler.go (hunk 20 above)
Comment (joechenrh) [reply pkg/dxf/export/scheduler.go]: Right, this was the root cause behind the P1 above — resolving through a session/domain only sees system tables in the scheduler's own keyspace. Fixed the same way: `snapshotTableInfos` now reads user tables straight through the meta API (`meta.NewReader(s.store.GetSnapshot(...))` + `GetDatabase`/`GetTable`) against the task's own store, no domain/infoschema involved at all.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 .

File: pkg/dxf/export/split.go (hunk 29 above)
Comment (joechenrh) [reply pkg/dxf/export/split.go]: Same fix as the scheduler.go comment above — `split.go` no longer touches infoschema at all now. `tableInfoByID`/`splitTable`/`generateChunks` take a plain `map[int64]*model.TableInfo` that the scheduler builds once via the meta API in `snapshotTableInfos` and passes down, so there's no system-ks-only schema loading in the split path.

Done in https://github.com/pingcap/tidb/pull/70153/commits/a376a1a3ce8cf3d3a3d7c7a16108ab3210592867 .
